---
title: "ATAC vs RNA"
author: "Yixin (Izzy) Zhang"
date: "2024-01-31"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Reading Data and Determining Distribution
gene of interest: SOX2 (ENSG00000181449)
area of interest: HES1
## vroom to get columns based on col name
try fetching colnames based on gene names

## try with vroom
```{r}
## get all row names (genes) from the data
library(vroom)

file_path <- "./GSE162170/GSE162170_multiome_rna_counts.tsv.gz"

columns_to_read <- 1

# Read the file with vroom, specifying columns to read
Sys.setenv("VROOM_CONNECTION_SIZE"=1e10)#avoid error
gene_names <- vroom(file_path, col_select = columns_to_read, col_names = FALSE)

head(gene_names)
```

```{r}
gene_of_interest = "ENSG00000181449"

# try fetching colnames based on gene names
match_position <- match(gene_of_interest, gene_names$X1)

print(match_position) 
print(gene_names[match_position,])

# Read the specific row
# skip = row_index - 1 because skip is 0-based
gene_row <- vroom(file_path, skip = match_position - 1, n_max = 1, col_names = FALSE)

gene_row
```
## summary stats and plot
```{r}
library(tibble)
sox2 <- as.data.frame(t(gene_row[,-1]))
summary(sox2$V1)

# Check for any missing values
sum(is.na(sox2$V1))

# Histogram and density plot
hist(sox2$V1, breaks=50, main="Distribution of Gene Expression", xlab="Expression Level", col="blue")
plot(density(sox2$V1, na.rm = TRUE), main="Density Plot of Gene Expression", xlab="Expression Level", ylab="Density")

# Variance-to-Mean Ratio (VMR) to assess overdispersion
vmr <- var(sox2$V1) / mean(sox2$V1)
print(paste("Variance-to-Mean Ratio (VMR):", vmr))

# test overdispersion
library(MASS)
glm_pois <- glm(V1 ~ 1, family = "poisson", data = sox2)
library(AER)
dispersion_test <- dispersiontest(glm_pois)
print(dispersion_test)
```

```{r}
# fit poisson
library(ggplot2)
# get mean
lambda <- mean(sox2$V1)
# get poisson distr
exp_levels <- 0:max(sox2$V1)
poisson_df <- data.frame(expression_level = exp_levels,
                         frequency = dpois(exp_levels, lambda))

# plot
ggplot(sox2, aes(x = V1)) +
  geom_histogram(aes(y = after_stat(density)), binwidth = 1, fill = "gray", color = "black", alpha = 0.5) +
  geom_line(data = poisson_df, aes(x = expression_level, y = frequency), color = "blue", size = 1.5) +
  labs(title = "Histogram of Gene Expression with Fitted Poisson Distribution",
       x = "Expression Level",
       y = "Density") +
  theme_minimal()

```

```{r}
# fit NB
nb_model <- glm.nb(V1 ~ 1, data = sox2)

size <- nb_model$theta # Dispersion parameter
mu <- exp(coef(nb_model)[1]) # Mean parameter

exp_levels <- 0:max(sox2$V1)
predicted_freq_nb <- dnbinom(exp_levels, size=size, mu=mu)
df_nb <- data.frame(expression_level = exp_levels, predicted_freq = predicted_freq_nb)

# Plotting
ggplot(sox2, aes(x = V1)) +
  geom_histogram(aes(y = after_stat(density)), binwidth = 1, fill = "gray", color = "black", alpha = 0.5) +
  geom_line(data = df_nb, aes(x = expression_level, y = predicted_freq), color = "red", size = 1.5) +
  labs(title = "Histogram of Gene Expression with Fitted Negative Binomial Model",
       x = "Expression Level",
       y = "Density") +
  theme_minimal()
```

```{r}
# Compare AIC values
aic_nb <- AIC(nb_model)
aic_pois <- AIC(glm_pois)

aic_nb
aic_pois
```

negative binomial fits better, backed up by both visualization and having smaller AIC. 

## ATACseq (probability of gene activity)
```{r}
library(vroom)

file_path <- "./GSE162170/GSE162170_multiome_atac_gene_activities.tsv.gz"

columns_to_read <- 1

# Read the file with vroom, specifying columns to read
Sys.setenv("VROOM_CONNECTION_SIZE"=1e10)#avoid error
area_names <- vroom(file_path, col_select = 1, col_names = FALSE)

head(area_names)
```

```{r}
area_of_interest = "HES1"

# try fetching colnames based on gene names
area_position <- match(area_of_interest, area_names$X1)

print(area_position) 
print(area_names[area_position,])

# Read the specific row
# Note: skip = row_index - 1 because skip is 0-based and we're adjusting for R's 1-based indexing
area_row <- vroom(file_path, skip = area_position - 1, n_max = 1, col_names = FALSE)

# Display the fetched row
print(area_row)
```

## summary stats and plot
```{r}
library(tibble)
hes1 <- as.data.frame(t(area_row[,-1]))
summary(hes1$V1)

# Check for any missing values
sum(is.na(hes1$V1))

# Histogram and density plot to visualize the distribution of expression levels
hist(hes1$V1, breaks=50, main="Distribution of Gene Expression", xlab="Expression Level", col="blue")
plot(density(hes1$V1, na.rm = TRUE), main="Density Plot of Gene Expression", xlab="Expression Level", ylab="Density")
```


```{r}
library(gamlss)
data <- hes1$V1
fit_beinf <- gamlss(data ~ 1, family = BEINF)
summary(fit_beinf)

predicted_probs <- predict(fit_beinf, type = "response")

ggplot(hes1, aes(x = V1)) +
  geom_histogram(aes(y = ..density..), fill = "gray", color = "black", alpha = 0.5) +
  geom_density(aes(y = ..density..), adjust = 1, color = "blue", size = 1) +
  labs(title = "Histogram of Actual Data with Fitted Zero-One-Inflated Beta Distribution",
       x = "Value",
       y = "Density") +
  theme_minimal()
``` 

# Fit Simple GLM
## prep data
```{r}
dat <- data.frame(unlist(sox2$V1), unlist(hes1$V1))
names(dat) <- c("RNA", "ATAC")
```
## predict ATAC (ZI-BETA) from RNA
```{r}
library(gamlss)
model_zib <- gamlss(ATAC ~ RNA, family=BEZI, data=dat, control=gamlss.control(n.cyc=500, trace = FALSE))
summary(model_zib)
```
## predict RNA (NB) from ATAC
```{r}
library(MASS)
model_nb <- glm.nb(RNA ~ ATAC, data=dat)
summary(model_nb)
```
## 5 fold cv
### ZI-BETA
```{r}
library(gamlss)
library(caret)
library(Metrics)

## 5-fold cross-validation
set.seed(123) 
folds <- createFolds(dat$ATAC, k=5)

cor_zib <- vector("list", length = 5)

for(i in seq_along(folds)) {
  training <- dat[-folds[[i]],]
  testing <- dat[folds[[i]],]
  
  model_zib <- gamlss(ATAC ~ RNA, family=BEZI, data=training, control=gamlss.control(n.cyc=500, trace = FALSE))
  predictions <- predict(model_zib, newdata = testing, type="response")
  cor_zib[[i]] <- cor(predictions, testing$ATAC)
}

# Calculate average performance
average_cor_zib <- mean(sapply(cor_zib, function(x) x))

average_cor_zib
```
### NB
```{r}
library(MASS)
library(caret)

set.seed(123) 
folds_nb <- createFolds(dat$RNA, k=5)

cor_nb <- vector("list", length = 5)

for(i in seq_along(folds_nb)) {
  training_nb <- dat[-folds_nb[[i]],]
  testing_nb <- dat[folds_nb[[i]],]
  
  model_nb <- glm.nb(RNA ~ ATAC, data=training_nb)
  predictions_nb <- predict(model_nb, newdata = testing_nb, type="response")
  cor_nb[[i]] <- cor(predictions_nb, testing_nb$ATAC)
}

average_cor_nb <- mean(sapply(cor_nb, function(x) x))

average_cor_nb
```

## Bayesian (try rstan; weakly informative prior)
```{r}
# library(rstan)
# 
# stan_data <- list(N = nrow(dat),
#                   RNA = dat$RNA,
#                   ATAC = dat$ATAC)
# 
# fit <- stan(file = 'rna_to_atac.stan', data = stan_data, iter = 2000, chains = 4)
# 
# print(fit)
```

