---
title: "ATAC vs RNA"
author: "Yixin (Izzy) Zhang"
date: "2024-01-31"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Reading Data and Determining Distribution
gene of interest: SOX2 (ENSG00000181449)
area of interest: HES1
## vroom to get columns based on col name
try fetching colnames based on gene names

## try with vroom
```{r}
## get all row names (genes) from the data
library(vroom)

rna_file_path <- "./GSE162170/GSE162170_multiome_rna_counts.tsv.gz"

columns_to_read <- 1

# Read the file with vroom, specifying columns to read
Sys.setenv("VROOM_CONNECTION_SIZE"=1e10)#avoid error
gene_names <- vroom(file_path, col_select = columns_to_read, col_names = FALSE)

head(gene_names)
```

```{r}
gene_of_interest = "ENSG00000181449"

# try fetching colnames based on gene names
match_position <- match(gene_of_interest, gene_names$X1)

print(match_position) 
print(gene_names[match_position,])

# Read the specific row
# skip = row_index - 1 because skip is 0-based
gene_row <- vroom(file_path, skip = match_position - 1, n_max = 1, col_names = FALSE)

gene_row
```
## summary stats and plot
```{r}
library(tibble)
sox2 <- as.data.frame(t(gene_row[,-1]))
summary(sox2$V1)

# Check for any missing values
sum(is.na(sox2$V1))

# Histogram and density plot
hist(sox2$V1, breaks=50, main="Distribution of Gene Expression", xlab="Expression Level", col="blue")
plot(density(sox2$V1, na.rm = TRUE), main="Density Plot of Gene Expression", xlab="Expression Level", ylab="Density")

# Variance-to-Mean Ratio (VMR) to assess overdispersion
vmr <- var(sox2$V1) / mean(sox2$V1)
print(paste("Variance-to-Mean Ratio (VMR):", vmr))

# test overdispersion
library(MASS)
glm_pois <- glm(V1 ~ 1, family = "poisson", data = sox2)
library(AER)
dispersion_test <- dispersiontest(glm_pois)
print(dispersion_test)
```

```{r}
# fit poisson
library(ggplot2)
# get mean
lambda <- mean(sox2$V1)
# get poisson distr
exp_levels <- 0:max(sox2$V1)
poisson_df <- data.frame(expression_level = exp_levels,
                         frequency = dpois(exp_levels, lambda))

# plot
ggplot(sox2, aes(x = V1)) +
  geom_histogram(aes(y = after_stat(density)), binwidth = 1, fill = "gray", color = "black", alpha = 0.5) +
  geom_line(data = poisson_df, aes(x = expression_level, y = frequency), color = "blue", size = 1.5) +
  labs(title = "Histogram of Gene Expression with Fitted Poisson Distribution",
       x = "Expression Level",
       y = "Density") +
  theme_minimal()

```

```{r}
# fit NB
nb_model <- glm.nb(V1 ~ 1, data = sox2)

size <- nb_model$theta # Dispersion parameter
mu <- exp(coef(nb_model)[1]) # Mean parameter

exp_levels <- 0:max(sox2$V1)
predicted_freq_nb <- dnbinom(exp_levels, size=size, mu=mu)
df_nb <- data.frame(expression_level = exp_levels, predicted_freq = predicted_freq_nb)

# Plotting
ggplot(sox2, aes(x = V1)) +
  geom_histogram(aes(y = after_stat(density)), binwidth = 1, fill = "gray", color = "black", alpha = 0.5) +
  geom_line(data = df_nb, aes(x = expression_level, y = predicted_freq), color = "red", size = 1.5) +
  labs(title = "Histogram of Gene Expression with Fitted Negative Binomial Model",
       x = "Expression Level",
       y = "Density") +
  theme_minimal()
```

```{r}
# Compare AIC values
aic_nb <- AIC(nb_model)
aic_pois <- AIC(glm_pois)

aic_nb
aic_pois
```

negative binomial fits better, backed up by both visualization and having smaller AIC. 

## ATACseq (probability of gene activity)
```{r}
library(vroom)

atac_act_file_path <- "./GSE162170/GSE162170_multiome_atac_gene_activities.tsv.gz"

columns_to_read <- 1

# Read the file with vroom, specifying columns to read
Sys.setenv("VROOM_CONNECTION_SIZE"=1e10)#avoid error
area_names <- vroom(atac_act_file_path, col_select = columns_to_read, col_names = FALSE)

head(area_names)
```

```{r}
area_of_interest = "SOX2"

# try fetching colnames based on gene names
area_position <- match(area_of_interest, area_names$X1)

print(area_position) 
print(area_names[area_position,])

# Read the specific row
# Note: skip = row_index - 1 because skip is 0-based and we're adjusting for R's 1-based indexing
area_row <- vroom(atac_act_file_path, skip = area_position - 1, n_max = 1, col_names = FALSE)

# Display the fetched row
print(area_row)
```

## summary stats and plot
```{r}
library(tibble)
atac <- as.data.frame(t(area_row[,-1]))
summary(atac$V1)

# Check for any missing values
sum(is.na(atac$V1))

# Histogram and density plot to visualize the distribution of expression levels
hist(atac$V1, breaks=50, main="Distribution of Gene Expression", xlab="Expression Level", col="blue")
plot(density(atac$V1, na.rm = TRUE), main="Density Plot of Gene Expression", xlab="Expression Level", ylab="Density")
```


```{r}
library(gamlss)
data <- atac$V1
fit_beinf <- gamlss(data ~ 1, family = BEINF)
summary(fit_beinf)

predicted_probs <- predict(fit_beinf, type = "response")

ggplot(atac, aes(x = V1)) +
  geom_histogram(aes(y = ..density..), fill = "gray", color = "black", alpha = 0.5) +
  geom_density(aes(y = ..density..), adjust = 1, color = "blue", size = 1) +
  labs(title = "Histogram of Actual Data with Fitted Zero-One-Inflated Beta Distribution",
       x = "Value",
       y = "Density") +
  theme_minimal()
``` 

# Fit Simple GLM
## prep data
```{r}
dat <- data.frame(unlist(sox2$V1), unlist(atac$V1))
names(dat) <- c("RNA", "ATAC")
```
## predict ATAC (ZI-BETA) from RNA
```{r}
library(gamlss)
model_zib <- gamlss(ATAC ~ RNA, family=BEZI, data=dat, control=gamlss.control(n.cyc=500, trace = FALSE))
summary(model_zib)
```
## predict RNA (NB) from ATAC
```{r}
library(MASS)
model_nb <- glm.nb(RNA ~ ATAC, data=dat)
summary(model_nb)
```
## 5 fold cv
### ZI-BETA
```{r}
library(gamlss)
library(caret)
library(Metrics)

## 5-fold cross-validation
set.seed(123) 
folds <- createFolds(dat$ATAC, k=5)

cor_zib <- vector("list", length = 5)

for(i in seq_along(folds)) {
  training <- dat[-folds[[i]],]
  testing <- dat[folds[[i]],]
  
  model_zib <- gamlss(ATAC ~ RNA, family=BEZI, data=training, control=gamlss.control(n.cyc=500, trace = FALSE))
  predictions <- predict(model_zib, newdata = testing, type="response")
  cor_zib[[i]] <- cor(predictions, testing$ATAC)
}

# Calculate average performance
average_cor_zib <- mean(sapply(cor_zib, function(x) x))

average_cor_zib
```
Not very good at predicting ATACseq. 
### NB
```{r}
library(MASS)
library(caret)

set.seed(123) 
folds_nb <- createFolds(dat$RNA, k=5)

cor_nb <- vector("list", length = 5)

for(i in seq_along(folds_nb)) {
  training_nb <- dat[-folds_nb[[i]],]
  testing_nb <- dat[folds_nb[[i]],]
  
  model_nb <- glm.nb(RNA ~ ATAC, data=training_nb)
  predictions_nb <- predict(model_nb, newdata = testing_nb, type="response")
  cor_nb[[i]] <- cor(predictions_nb, testing_nb$RNA)
}

average_cor_nb <- mean(sapply(cor_nb, function(x) x))

average_cor_nb
```
Not good. 

## Bayesian (try rstan; weakly informative prior)
```{r}
# library(rstan)
# 
# stan_data <- list(N = nrow(dat),
#                   RNA = dat$RNA,
#                   ATAC = dat$ATAC)
# 
# fit <- stan(file = 'rna_to_atac.stan', data = stan_data, iter = 2000, chains = 4)
# 
# print(fit)
```

## getting more variables (SOX2, HES1, PAX6, SOX6, SOX9, GLI3, ASCL1)

```{r}
# multiple_atac_of_interest = c("SOX2", "HES1", "PAX6", "SOX6", "SOX9", "GLI3", "ASCL1")
# 
# # try fetching colnames based on gene names
# multiple_atac_positions <- match(multiple_atac_of_interest, area_names$X1)
# 
# print(multiple_atac_positions) 
# print(area_names[multiple_atac_positions,])

```

```{r}
# sox2_atac_tmp <- vroom(atac_act_file_path, skip = 4234 - 1, n_max = 1, col_names = FALSE)
# sox2_atac <- as.data.frame(t(sox2_atac_tmp[,-1]))
# hes1_atac_tmp <- vroom(atac_act_file_path, skip = 4318 - 1, n_max = 1, col_names = FALSE)
# hes1_atac <- as.data.frame(t(hes1_atac_tmp[,-1]))
# pax6_atac_tmp <- vroom(atac_act_file_path, skip = 10423 - 1, n_max = 1, col_names = FALSE)
# pax6_atac <- as.data.frame(t(pax6_atac_tmp[,-1]))
# sox6_atac_tmp <- vroom(atac_act_file_path, skip = 10352 - 1, n_max = 1, col_names = FALSE)
# sox6_atac <- as.data.frame(t(sox6_atac_tmp[,-1]))
# sox9_atac_tmp <- vroom(atac_act_file_path, skip = 15771 - 1, n_max = 1, col_names = FALSE)
# sox9_atac <- as.data.frame(t(sox9_atac_tmp[,-1]))
# gli3_atac_tmp <- vroom(atac_act_file_path, skip = 7244 - 1, n_max = 1, col_names = FALSE)
# gli3_atac <- as.data.frame(t(gli3_atac_tmp[,-1]))
# ascl1_atac_tmp <- vroom(atac_act_file_path, skip = 12147 - 1, n_max = 1, col_names = FALSE)
# ascl1_atac <- as.data.frame(t(ascl1_atac_tmp[,-1]))
# for (i in multiple_atac_positions){
#   area_row <- vroom(atac_act_file_path, skip = i - 1, n_max = 1, col_names = FALSE)
#   print("done")
#   atac <- as.data.frame(t(area_row[,-1]))
#   if (!exists("atac_df")){
#     atac_df <- atac
#   }else{atac_df <- cbind(atac_df, atac)}
#   print("done2")
# }


# Display the fetched row
# print(atac_df)
```

```{r}
# atac_df <- cbind(sox2_atac, hes1_atac, pax6_atac, sox6_atac, sox9_atac, gli3_atac, ascl1_atac)
# names(atac_df) <- c("SOX2", "HES1", "PAX6", "SOX6", "SOX9", "GLI3", "ASCL1")
# write.csv(atac_df, file = "./test_data/atac_7_var.csv", row.names = FALSE)
```

## getting more rna variables
```{r}
# multiple_rna_of_interest = c("ENSG00000181449", "ENSG00000114315", "ENSG00000007372", "ENSG00000110693", "ENSG00000125398", "ENSG00000106571", "ENSG00000139352")
# 
# # try fetching colnames based on gene names
# multiple_rna_positions <- match(multiple_rna_of_interest, gene_names$X1)
# 
# print(multiple_rna_positions) 
# print(gene_names[multiple_rna_positions,])
```

```{r}
# for (i in multiple_rna_positions){
#   area_row <- vroom(rna_file_path, skip = i - 1, n_max = 1, col_names = FALSE)
#   print("done")
#   rna <- as.data.frame(t(area_row[,-1]))
#   if (!exists("rna_df")){
#     rna_df <- rna
#   }else{rna_df <- cbind(rna_df, rna)}
#   print("done2")
# }
# 
# 
# # Display the fetched row
# print(rna_df)
```
```{r}
# names(rna_df) <- c("SOX2", "HES1", "PAX6", "SOX6", "SOX9", "GLI3", "ASCL1")
# write.csv(rna_df, file = "./test_data/rna_7_var.csv", row.names = FALSE)
```

```{r}
names(atac_df) <- c("SOX2_ATAC", "HES1_ATAC", "PAX6_ATAC", "SOX6_ATAC", "SOX9_ATAC", "GLI3_ATAC", "ASCL1_ATAC")
names(rna_df) <- c("SOX2_RNA", "HES1_RNA", "PAX6_RNA", "SOX6_RNA", "SOX9_RNA", "GLI3_RNA", "ASCL1_RNA")
total_df <- cbind(atac_df, rna_df)
write.csv(total_df, file = "./test_data/total_7_var.csv", row.names = FALSE)
```

## fit NB and ZIBeta for data with 7 variables
```{r}
total_multi_var <- read.csv("./test_data/total_7_var.csv", header = T)
```
### predict ATAC (ZI-BETA) from RNA
```{r}
library(gamlss)
model_zib_multi_var <- gamlss(SOX2_ATAC ~ SOX2_RNA + HES1_RNA + PAX6_RNA + SOX6_RNA + SOX9_RNA + GLI3_RNA + ASCL1_RNA, family=BEZI, data=total_multi_var, control=gamlss.control(n.cyc=500, trace = FALSE))
summary(model_zib_multi_var)
```
### predict RNA (NB) from ATAC
```{r}
library(MASS)
model_nb_multi_var <- glm.nb(SOX2_RNA ~ SOX2_ATAC + HES1_ATAC + PAX6_ATAC + SOX6_ATAC + SOX9_ATAC + GLI3_ATAC + ASCL1_ATAC, data=total_multi_var)
summary(model_nb_multi_var)
```
### 5 fold cv
#### ZI-Beta (predict ATAC from RNA)
```{r}
library(gamlss)
library(caret)
library(Metrics)

## 5-fold cross-validation
set.seed(123) 
folds <- createFolds(total_multi_var$SOX2_ATAC, k=5)

cor_zib <- vector("list", length = 5)

for(i in seq_along(folds)) {
  training <- total_multi_var[-folds[[i]],]
  testing <- total_multi_var[folds[[i]],]
  
  model_zib <- gamlss(SOX2_ATAC ~ SOX2_RNA + HES1_RNA + PAX6_RNA + SOX6_RNA + SOX9_RNA + GLI3_RNA + ASCL1_RNA, family=BEZI, data=training, control=gamlss.control(n.cyc=500, trace = FALSE))
  predictions <- predict(model_zib, newdata = testing, type="response")
  cor_zib[[i]] <- cor(predictions, testing$SOX2_ATAC)
}

# Calculate average performance
average_cor_zib <- mean(sapply(cor_zib, function(x) x))

average_cor_zib
```
some improvement from just SOX2, but still not good. 
#### NB (predict RNA from ATAC)
```{r}
library(MASS)
library(caret)

set.seed(123) 
folds_nb <- createFolds(total_multi_var$SOX2_RNA, k=5)

cor_nb <- vector("list", length = 5)

for(i in seq_along(folds_nb)) {
  training_nb <- total_multi_var[-folds_nb[[i]],]
  testing_nb <- total_multi_var[folds_nb[[i]],]
  
  model_nb <- glm.nb(SOX2_RNA ~ SOX2_ATAC + HES1_ATAC + PAX6_ATAC + SOX6_ATAC + SOX9_ATAC + GLI3_ATAC + ASCL1_ATAC, data=training_nb)
  predictions_nb <- predict(model_nb, newdata = testing_nb, type="response")
  cor_nb[[i]] <- cor(predictions_nb, testing_nb$SOX2_RNA)
}

average_cor_nb <- mean(sapply(cor_nb, function(x) x))

average_cor_nb
```
some improvement, but not good. 

Remove SOX6_ATAC since its coefficient was not significant in previous fit without cv. 
```{r}
library(MASS)
library(caret)

set.seed(123) 
folds_nb <- createFolds(total_multi_var$SOX2_RNA, k=5)

cor_nb <- vector("list", length = 5)

for(i in seq_along(folds_nb)) {
  training_nb <- total_multi_var[-folds_nb[[i]],]
  testing_nb <- total_multi_var[folds_nb[[i]],]
  
  model_nb <- glm.nb(SOX2_RNA ~ SOX2_ATAC + HES1_ATAC + PAX6_ATAC + SOX9_ATAC + GLI3_ATAC + ASCL1_ATAC, data=training_nb)
  predictions_nb <- predict(model_nb, newdata = testing_nb, type="response")
  cor_nb[[i]] <- cor(predictions_nb, testing_nb$SOX2_RNA)
}

average_cor_nb <- mean(sapply(cor_nb, function(x) x))

average_cor_nb
```

```{r}
vif_results <- vif(model_nb_multi_var)
print(vif_results)
```


### zero inflated NB (desparate try, did not varify if distribution of RNA is zero-inflated or not)
```{r}
library(pscl)

# Fitting a Zero-Inflated Negative Binomial Model
model_zinb <- zeroinfl(SOX2_RNA ~ SOX2_ATAC + HES1_ATAC + PAX6_ATAC + SOX6_ATAC + SOX9_ATAC + GLI3_ATAC + ASCL1_ATAC | 1, 
                       data = total_multi_var, 
                       dist = "negbin")

# Summary of the fitted model
summary(model_zinb)

```

```{r}
library(MASS)
library(caret)

set.seed(123) 
folds_nb <- createFolds(total_multi_var$SOX2_RNA, k=5)

cor_nb <- vector("list", length = 5)

for(i in seq_along(folds_nb)) {
  training_nb <- total_multi_var[-folds_nb[[i]],]
  testing_nb <- total_multi_var[folds_nb[[i]],]
  
  model_zinb <- zeroinfl(SOX2_RNA ~ SOX2_ATAC + HES1_ATAC + PAX6_ATAC + SOX6_ATAC + SOX9_ATAC + GLI3_ATAC + ASCL1_ATAC | 1, 
                       data = total_multi_var, 
                       dist = "negbin")

  predictions_zinb <- predict(model_zinb, newdata = testing_nb, type="response")
  cor_nb[[i]] <- cor(predictions_zinb, testing_nb$SOX2_RNA)
}

average_cor_nb <- mean(sapply(cor_nb, function(x) x))

average_cor_nb
```
# plot the relationship between SOX2 ATACseq and SOX2 RNAseq
```{r}
library(ggplot2)

# Basic scatter plot
ggplot(total_multi_var, aes(x = SOX2_ATAC, y = SOX2_RNA)) +
  geom_point() +
  labs(x = "SOX2 ATAC", y = "SOX2 RNA", title = "Relationship between SOX2 ATAC and SOX2 RNA")

# Log scale plot, if needed
ggplot(total_multi_var, aes(x = SOX2_ATAC, y = SOX2_RNA)) +
  geom_point() +
  scale_x_log10() +
  scale_y_log10() +
  labs(x = "Log(SOX2 ATAC)", y = "Log(SOX2 RNA)", title = "Log-Scale Relationship between SOX2 ATAC and SOX2 RNA")

```

# fit GAM
From results of GLM and plot, we see that the relationships are non-linear
## GAM, RNA~ATAC
### RNA~1 ATAC var
```{r}
library(mgcv)
model_nb_gam_single_var <- gam(SOX2_RNA ~ s(SOX2_ATAC), data=total_multi_var)
summary(model_nb_gam_single_var)
```

```{r}
library(mgcv)
library(caret)

set.seed(123)
folds_nb <- createFolds(total_multi_var$SOX2_RNA, k=5)

cor_nb_gam <- vector("list", length = 5)

for(i in seq_along(folds_nb)) {
  training_nb <- total_multi_var[-folds_nb[[i]], ]
  testing_nb <- total_multi_var[folds_nb[[i]], ]
  
  # Fitting a GAM model
  model_gam <- gam(SOX2_RNA ~ s(SOX2_ATAC), data=training_nb)
  predictions_gam <- predict(model_gam, newdata=testing_nb, type="response")
  cor_nb_gam[[i]] <- cor(predictions_gam, testing_nb$SOX2_RNA)
}

average_cor_nb_gam <- mean(sapply(cor_nb_gam, function(x) x))

average_cor_nb_gam
```
### RNA~7 ATAC vars
```{r}
library(mgcv)
model_rna_gam_multi_var <- gam(SOX2_RNA ~ s(SOX2_ATAC) + s(HES1_ATAC) + s(PAX6_ATAC) + s(SOX6_ATAC) + s(SOX9_ATAC) + s(GLI3_ATAC) + s(ASCL1_ATAC), data=total_multi_var)
summary(model_rna_gam_multi_var)
```

```{r}
library(mgcv)
library(caret)

set.seed(123)
folds_nb <- createFolds(total_multi_var$SOX2_RNA, k=5)

cor_nb_gam <- vector("list", length = 5)

for(i in seq_along(folds_nb)) {
  training_nb <- total_multi_var[-folds_nb[[i]], ]
  testing_nb <- total_multi_var[folds_nb[[i]], ]
  
  # Fitting a GAM model
  model_gam <- gam(SOX2_RNA ~ s(SOX2_ATAC) + s(HES1_ATAC) + s(PAX6_ATAC) + s(SOX6_ATAC) + s(SOX9_ATAC) + s(GLI3_ATAC) + s(ASCL1_ATAC), data=training_nb)
  predictions_gam <- predict(model_gam, newdata=testing_nb, type="response")
  cor_nb_gam[[i]] <- cor(predictions_gam, testing_nb$SOX2_RNA)
}

average_cor_nb_gam <- mean(sapply(cor_nb_gam, function(x) x))

average_cor_nb_gam

```

#### plot fitted and actual val
```{r}
library(mgcv)
library(ggplot2)

# Generate predictions on the same data used for fitting the model for demonstration
predicted_SOX2_rna <- predict(model_rna_gam_multi_var, type = "response")

# Create a histogram of the true values
p <- ggplot(total_multi_var, aes(x = SOX2_RNA)) +
  geom_histogram(aes(y = ..density..), binwidth = diff(range(total_multi_var$SOX2_RNA))/30, colour = "black", fill = "white") +
  labs(x = "SOX2_RNA", y = "Density") +
  ggtitle("True vs. Predicted Distribution of SOX2_RNA") +
  theme_minimal()

# Overlay the predicted value distribution as a density plot
p + geom_density(aes(x = predicted_SOX2_rna), colour = "blue")

```
predicted values are continuous, **round down** to nearest number and plot again. (tried in cross validation, correlation is 0.35, not improve much) (round to nearest number gave lower cor = 0.32)

```{r}
library(mgcv)
library(ggplot2)

# Generate predictions on the same data used for fitting the model for demonstration
predicted_SOX2_rna <- predict(model_rna_gam_multi_var, type = "response")
predicted_SOX2_rna_round <- round(predicted_SOX2_rna)
predicted_SOX2_rna_floor <- floor(predicted_SOX2_rna)


# Create a histogram of the true values
p <- ggplot(total_multi_var, aes(x = SOX2_RNA)) +
  geom_histogram(aes(y = ..density..), binwidth = diff(range(total_multi_var$SOX2_RNA))/30, colour = "black", fill = "white") +
  labs(x = "SOX2_RNA", y = "Density") +
  ggtitle("True vs. Predicted Distribution of SOX2_RNA") +
  theme_minimal()

# Overlay the predicted value distribution as a density plot
p + geom_density(aes(x = predicted_SOX2_rna), colour = "blue") + 
  geom_density(aes(x = predicted_SOX2_rna_round), colour = "green") + 
  geom_density(aes(x = predicted_SOX2_rna_floor), colour = "red")

plot(p)
```
```{r}
plot(predicted_SOX2_rna, total_multi_var$SOX2_RNA)
plot(log(predicted_SOX2_rna), log(total_multi_var$SOX2_RNA))
```

## GAM, ATAC~RNA
### ATAC~1 RNA var
```{r}
library(mgcv)
model_atac_gam_single_var <- gam(SOX2_ATAC ~ s(SOX2_RNA), data=total_multi_var)
summary(model_atac_gam_single_var)
```

```{r}
library(mgcv)
library(caret)

set.seed(123)
folds_atac <- createFolds(total_multi_var$SOX2_ATAC, k=5)

cor_atac_gam <- vector("list", length = 5)

for(i in seq_along(folds_nb)) {
  training_atac <- total_multi_var[-folds_atac[[i]], ]
  testing_atac <- total_multi_var[folds_atac[[i]], ]
  
  # Fitting a GAM model
  model_gam <- gam(SOX2_ATAC ~ s(SOX2_RNA), data=training_atac)
  predictions_gam <- predict(model_gam, newdata=testing_atac, type="response")
  cor_atac_gam[[i]] <- cor(predictions_gam, testing_atac$SOX2_ATAC)
}

average_cor_atac_gam <- mean(sapply(cor_atac_gam, function(x) x))

average_cor_atac_gam
```

### RNA~7 ATAC vars
```{r}
library(mgcv)
model_atac_gam_multi_var <- gam(SOX2_ATAC ~ s(SOX2_RNA) + s(HES1_RNA) + s(PAX6_RNA) + s(SOX6_RNA) + s(SOX9_RNA) + s(GLI3_RNA) + s(ASCL1_RNA, k = 5), data=total_multi_var) ## + s(ASCL1_RNA) has too few unique values (0,1,...,7), therefore added k=5
summary(model_atac_gam_multi_var)
```
```{r}
library(mgcv)
library(caret)

set.seed(123)
folds_atac <- createFolds(total_multi_var$SOX2_ATAC, k=5)

cor_atac_gam <- vector("list", length = 5)

for(i in seq_along(folds_nb)) {
  training_atac <- total_multi_var[-folds_atac[[i]], ]
  testing_atac <- total_multi_var[folds_atac[[i]], ]
  
  # Fitting a GAM model
  model_gam <- gam(SOX2_ATAC ~ s(SOX2_RNA) + s(HES1_RNA) + s(PAX6_RNA) + s(SOX6_RNA) + s(SOX9_RNA) + s(GLI3_RNA) + s(ASCL1_RNA, k = 5), data=training_atac)
  predictions_gam <- predict(model_gam, newdata=testing_atac, type="response")
  cor_atac_gam[[i]] <- cor(predictions_gam, testing_atac$SOX2_ATAC)
}

average_cor_atac_gam <- mean(sapply(cor_atac_gam, function(x) x))

average_cor_atac_gam
```
#### plot fitted and actual val
```{r}
library(mgcv)
library(ggplot2)

# Assuming model_atac_gam_multi_var is already fitted as per your code
# Generate predictions on the same data used for fitting the model for demonstration
predicted_SOX2_atac <- predict(model_atac_gam_multi_var, type = "response")

# Create a histogram of the true values
p <- ggplot(total_multi_var, aes(x = SOX2_ATAC)) +
  geom_histogram(aes(y = ..density..), binwidth = diff(range(total_multi_var$SOX2_ATAC))/30, colour = "black", fill = "white") +
  labs(x = "SOX2_ATAC", y = "Density") +
  ggtitle("True vs. Predicted Distribution of SOX2_ATAC") +
  theme_minimal()

# Overlay the predicted value distribution as a density plot
p + geom_density(aes(x = predicted_SOX2_atac), colour = "blue")

```
#### fitted vs response
```{r}
plot(predicted_SOX2_atac, total_multi_var$SOX2_ATAC)
plot(log(predicted_SOX2_atac), log(total_multi_var$SOX2_ATAC))
```

