---
title: "EDA"
author: "Yixin (Izzy) Zhang"
date: "2024-01-31"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Reading Data
## read subset of data

```{r}
count_lines <- function(file_path) {
    line_count <- 0
    con <- file(file_path, open = "rt")
    while (length(readLines(con, n = 1)) > 0) {
        line_count <- line_count + 1
    }
    close(con)
    return(line_count)
}

file_path <- "./GSE162170/GSE162170_multiome_spliced_rna_counts.tsv.gz"  # Adjust this to your file path
total_lines <- count_lines(file_path)
```

```{r}
set.seed(123)  # For reproducibility
sample_size <- total_lines / 10
selected_lines <- sort(sample(total_lines, sample_size))
```

```{r}
read_selected_lines <- function(file_path, selected_lines) {
    con <- file(file_path, open = "rt")
    data <- NULL
    line_index <- 0
    while (length(line <- readLines(con, n = 1)) > 0) {
        line_index <- line_index + 1
        if (line_index %in% selected_lines) {
            data <- rbind(data, read.csv(text = line, sep = "\t", header = FALSE))
        }
    }
    close(con)
    return(data)
}

sampled_data <- read_selected_lines(file_path, selected_lines)
colnames(sampled_data) <- names(read.csv(file_path, nrows = 1, sep = "\t"))  # Set column names
```
```{r}
# Viewing the structure of the dataset
str(sampled_data)

# Basic summary statistics
summary(sampled_data)
```


## BPCells iterate
```{r}
library(BPCells)
suppressPackageStartupMessages({
  library(dplyr)
  library(patchwork)
})
```

convert RNA matrix to bitpacked compressed format
```{r}
setwd("./GSE162170")
# Check if we already ran import
if (!file.exists("rna_raw")) {
  mat_raw <- open_matrix_10x_hdf5("GSE162170_multiome_rna_counts.tsv.gz", feature_type="Gene Expression") %>% 
    write_matrix_dir("rna_raw")
} else {
  mat_raw <- open_matrix_dir("rna_raw")
}
mat_raw
```

## fread to get columns based on col name
gene of interest: SOX2 (ENSG00000181449)
area of interest: HES1
getting the row names (list of gene names)
```{r}
library(vroom)
Sys.setenv("VROOM_CONNECTION_SIZE"=1e10)

# Specify the path to your .tsv.gz file
file_path <- "./GSE162170/GSE162170_multiome_rna_counts.tsv.gz"

# Specify the columns you want to read by name or index
# For example, to read the first five columns, you can specify their indices as 1:5
# If you know the column names, you can specify them directly
columns_to_read <- 1

# Read the file with vroom, specifying columns to read
gene_names <- vroom(file_path, col_select = columns_to_read, col_names = FALSE)

# Check the first few rows of the loaded data subset
gene_names

```

<!-- (getting col names) -->
<!-- ```{r} -->
<!-- library(vroom) -->

<!-- # Define the path to your .tsv.gz file -->
<!-- file_path <- "./GSE162170/GSE162170_multiome_rna_counts.tsv.gz" -->

<!-- # Read in only the first few rows to get column names -->
<!-- data_preview <- vroom(file_path, n_max = 3) -->

<!-- # Display column names -->
<!-- colnames(data_preview) -->
<!-- ``` -->

<!-- fread to extract 1 column (Note: 1 placeholder is automatically placed at [1,1] since no colname for index) -->
<!-- ```{r} -->
<!-- library(data.table) -->


<!-- # Specify the columns you want to extract by name or by index -->
<!-- # For example, if you know the names of the columns -->
<!-- columns_of_interest <- c(1, 2) -->

<!-- # Use fread to read only the specified columns -->
<!-- data <- fread(file = file_path, select = columns_of_interest) -->

<!-- # Now `data` contains only the columns you're interested in -->

<!-- ``` -->

try fetching colnames based on gene names
```{r}
# Find the position of the string in the list
match_position <- match(gene_of_interest, gene_names$X1)

print(match_position) 
print(gene_names[match_position,])
```

```{r}
# Row index you want to fetch (e.g., the 10th row)
row_index <- 10

# Read the specific row
# Note: skip = row_index - 1 because skip is 0-based and we're adjusting for R's 1-based indexing
specific_row <- vroom(file_path, skip = match_position - 1, n_max = 1, col_names = FALSE)

# Display the fetched row
print(specific_row)

```

## try with vroom
```{r}
library(vroom)

file_path <- "./GSE162170/GSE162170_multiome_rna_counts.tsv.gz"

columns_to_read <- 1

# Read the file with vroom, specifying columns to read
Sys.setenv("VROOM_CONNECTION_SIZE"=1e10)#avoid error
gene_names <- vroom(file_path, col_select = columns_to_read, col_names = FALSE)

head(gene_names)
```

```{r}
gene_of_interest = "ENSG00000181449"

# try fetching colnames based on gene names
match_position <- match(gene_of_interest, gene_names$X1)

print(match_position) 
print(gene_names[match_position,])

# Read the specific row
# Note: skip = row_index - 1 because skip is 0-based and we're adjusting for R's 1-based indexing
specific_row <- vroom(file_path, skip = match_position - 1, n_max = 1, col_names = FALSE)

# Display the fetched row
print(specific_row)
```
## summary stats and plot
```{r}
library(tibble)
sox2 <- as.data.frame(t(specific_row[,-1]))
summary(sox2$V1)

# Check for any missing values
sum(is.na(sox2$V1))

# Histogram and density plot to visualize the distribution of expression levels
hist(sox2$V1, breaks=50, main="Distribution of Gene Expression", xlab="Expression Level", col="blue")
plot(density(sox2$V1, na.rm = TRUE), main="Density Plot of Gene Expression", xlab="Expression Level", ylab="Density")

# Variance-to-Mean Ratio (VMR) to assess overdispersion
vmr <- var(sox2$V1) / mean(sox2$V1)
print(paste("Variance-to-Mean Ratio (VMR):", vmr))

# test overdispersion
library(MASS)
glm_pois <- glm(V1 ~ 1, family = "poisson", data = sox2)
library(AER)
dispersion_test <- dispersiontest(glm_pois)
print(dispersion_test)
```
```{r}
library(ggplot2)

# Calculate the observed mean to use as lambda for the Poisson distribution
lambda <- mean(sox2$V1)

# Create a data frame for the Poisson distribution overlay
exp_levels <- 0:max(sox2$V1, na.rm = TRUE)
poisson_df <- data.frame(expression_level = exp_levels,
                         frequency = dpois(exp_levels, lambda))

# Plotting the histogram of observed data and overlaying the Poisson distribution
ggplot(sox2, aes(x = V1)) +
  geom_histogram(aes(y = after_stat(density)), binwidth = 1, fill = "gray", color = "black", alpha = 0.5) +
  geom_line(data = poisson_df, aes(x = expression_level, y = frequency), color = "blue", size = 1.5) +
  labs(title = "Histogram of Gene Expression with Fitted Poisson Distribution",
       x = "Expression Level",
       y = "Density") +
  theme_minimal()

```

```{r}
nb_model <- glm.nb(V1 ~ 1, data = sox2)

# Extract model parameters
size <- nb_model$theta # Dispersion parameter
mu <- exp(coef(nb_model)[1]) # Mean parameter

predicted_freq_nb <- dnbinom(exp_levels, size=size, mu=mu)
df_nb <- data.frame(expression_level = exp_levels, predicted_freq = predicted_freq_nb)

# Plotting
ggplot(sox2, aes(x = V1)) +
  geom_histogram(aes(y = ..density..), binwidth = 1, fill = "gray", color = "black", alpha = 0.5) +
  geom_line(data = df_nb, aes(x = expression_level, y = predicted_freq), color = "red", size = 1.5) +
  labs(title = "Histogram of Gene Expression with Fitted Negative Binomial Model",
       x = "Expression Level",
       y = "Density") +
  theme_minimal()
```

```{r}
# Compare AIC values
aic_nb <- AIC(nb_model)
aic_pois <- AIC(glm_pois)

aic_nb
aic_pois
```

negative binomial fits better, backed up by both visualization and having smaller AIC. 

## ATACseq
```{r}
library(vroom)

file_path <- "./GSE162170/GSE162170_multiome_atac_counts.tsv.gz"

columns_to_read <- 1

# Read the file with vroom, specifying columns to read
Sys.setenv("VROOM_CONNECTION_SIZE"=1e10)#avoid error
area_names <- vroom(file_path, col_select = 3, n_max = 1e3, col_names = FALSE)

head(area_names)
```

